src/proto/uprotocol/uresource.rs

```diff
@@ -0,0 +1,57 @@
+/********************************************************************************
+ * Copyright (c) 2023 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+use crate::uprotocol::UResource;
+
+impl UResource {
+    pub fn has_id(resource: &UResource) -> bool {
+        resource.id.is_some()
+    }
+
+    pub fn get_message(resource: &UResource) -> Option<&String> {
+        match &resource.message {
+            Some(message) => Some(message),
+            _ => None,
+        }
+    }
+
+    pub fn get_instance(resource: &UResource) -> Option<&String> {
+        match &resource.instance {
+            Some(instance) => Some(instance),
+            _ => None,
+        }
+    }
+}
+
+impl From<String> for UResource {
```

Comment:
A lot of allocations happen here. 
I suggest to implement the `From<&str>` and not to collect the split into a `Vec<_>`.
Using iterators you can achieve the same result with the difference of allocating just the fields necessary to construct `UResource` (and not any intermediary step).